hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ContainerReplicaPendingOps.java

```diff
@@ -0,0 +1,227 @@
+   * @param containerID ContainerID for which to remove the op.
+   * @param op ContainerReplicaOp to remove
+   * @return True if an element was found and deleted, false otherwise.
+   */
+  public boolean removeOp(ContainerID containerID,
+      ContainerReplicaOp op) {
+    return completeOp(op.getOpType(), containerID, op.getTarget(),
+        op.getReplicaIndex());
+  }
+
+  /**
+   * Iterate over all pending entries and remove any which have expired, meaning
+   * they have not completed the operation inside the given time.
+   * @param expiryMilliSeconds
+   */
+  public void removeExpiredEntries(long expiryMilliSeconds) {
+    for (ContainerID containerID : pendingOps.keySet()) {
+      // Rather than use an entry set, we get the map entry again. This is
+      // to protect against another thread modifying the value after this
+      // iterator started. Once we lock on the ContainerID object, no other
+      // changes can occur to the list of ops associated with it.
+      Lock lock = writeLock(containerID);
+      lock.lock();
+      try {
+        List<ContainerReplicaOp> ops = pendingOps.get(containerID);
+        if (ops == null) {
+          // There should not be null entries, but another thread may have
+          // removed the map entry after the iterator was started.
+          continue;
+        }
+        Iterator<ContainerReplicaOp> iterator = ops.listIterator();
+        while (iterator.hasNext()) {
+          ContainerReplicaOp op = iterator.next();
+          if (op.getScheduledEpochMillis() + expiryMilliSeconds
+              < clock.millis()) {
+            iterator.remove();
+          }
+        }
+        if (ops.size() == 0) {
+          pendingOps.remove(containerID);
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+  }
+
+  private void addReplica(ContainerReplicaOp.PendingOpType opType,
+      ContainerID containerID, DatanodeDetails target, int replicaIndex) {
+    Lock lock = writeLock(containerID);
+    lock.lock();
+    try {
+      List<ContainerReplicaOp> ops = pendingOps.computeIfAbsent(
+          containerID, s -> new ArrayList<>());
+      ops.add(new ContainerReplicaOp(opType,
+          target, replicaIndex, clock.millis()));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private boolean completeOp(ContainerReplicaOp.PendingOpType opType,
+      ContainerID containerID, DatanodeDetails target, int replicaIndex) {
+    boolean found = false;
+    Lock lock = writeLock(containerID);
+    lock.lock();
+    try {
+      List<ContainerReplicaOp> ops = pendingOps.get(containerID);
+      if (ops != null) {
+        Iterator<ContainerReplicaOp> iterator = ops.listIterator();
+        while (iterator.hasNext()) {
+          ContainerReplicaOp op = iterator.next();
+          if (op.getOpType() == opType
+              && op.getTarget().equals(target)
+              && op.getReplicaIndex() == replicaIndex) {
+            found = true;
```

Comment:
NIT: can we add rewrite `equals` function for `ContainerReplicaOp`?

Reply:
I did think about adding an equals method to ContainerReplicaPendingOp, however I was not sure whether to include the timestamp in it or not.

Also, we compare the fields differently in a couple of places, so the equals will not help there, but it would make this method a bit nicer. I will look into adding it.

Reply:
I started the add the equals method, and I don't think it makes the code better. First its questionable whether timestamp should be included in it. Even if I adjusted the code in this method to use `equals`, I would need to form a new ContainerReplicaOp object from the parameters passed into complete methods, which doesn't seem necessary. I feel it may be better to leave it as it is now. What do you think?

Reply:
agree, it makes sense